### PR TITLE
bot: build detailed reports for tests suites & platforms

### PR DESCRIPTION
**Pull request by La0**

### CODE REVIEW

File: bot/code_coverage_bot/artifacts.py [hunk 1]
```diff
@@ -25,41 +30,65 @@ def __init__(self, task_ids, parent_dir="ccov-artifacts", task_name_filter="*"):
...
-            for f in os.listdir(self.parent_dir)
-            if os.path.basename(f).startswith(f"{platform}_")
+            artifact.chunk
+            for artifact in self.artifacts
+            if artifact.platform == platform
         )
 
-    def get(self, platform=None, suite=None, chunk=None):
-        files = os.listdir(self.parent_dir)
+    def get_suites(self):
```
Comment (marco-c) [inline bot/code_coverage_bot/artifacts.py]: Nit: get_suite_platform_combinations?

File: bot/tests/test_artifacts.py [hunk 2]
```diff
@@ -50,11 +66,56 @@ def test_get_chunks(FAKE_ARTIFACTS_DIR):
...
+            ["linux_firefox-ui-functional-remote_code-coverage-jsvm.info"]
+        ),
+        ("linux", "firefox-ui-functional"): add_dir(
+            ["linux_firefox-ui-functional-remote_code-coverage-jsvm.info"]
+        ),
+        ("all", "mochitest"): add_dir(
+            [
+                "windows_mochitest-1_code-coverage-jsvm.info",
+                "linux_mochitest-2_code-coverage-grcov.zip",
+            ]
```
Comment (marco-c) [inline bot/tests/test_artifacts.py]: Will the order always be the same here?

File: bot/code_coverage_bot/uploader.py [hunk 3]
```diff
@@ -47,41 +47,45 @@ def gcp(repository, revision, report, platform=None, suite=None):
     logger.info("Uploaded {} on {}".format(path, bucket))
 
     # Trigger ingestion on backend
```
Comment (marco-c) [inline bot/code_coverage_bot/uploader.py]: Do we want to ingest all platform/suite combinations? Or should we do it on demand only when somebody requests? (I guess they will be requested seldomly compared to the full report)

File: bot/code_coverage_bot/codecov.py [hunk 4]
```diff
@@ -185,7 +187,25 @@ def go_from_trigger_mozilla_central(self):
...
+    def go_from_trigger_mozilla_central(self):
+        # Check the covdir report does not already exists
+        if uploader.gcp_covdir_exists(self.branch, self.revision, "all", "all"):
+            logger.warn("Full covdir report already on GCP")
+            return
+
+        self.retrieve_source_and_artifacts()
+
+        # TODO: restore that check
+        # self.check_javascript_files()
```
Comment (marco-c) [inline bot/code_coverage_bot/codecov.py]: Need to restore the check

File: bot/code_coverage_bot/codecov.py [hunk 5]
```diff
@@ -146,27 +142,33 @@ def build_suites(self):
...
-        # If they don't, there might be a bug in the LCOV rewriter.
+    def check_javascript_files(self):
+        """
+        Check that all JavaScript files present in the coverage artifacts actually exist.
+        If they don't, there might be a bug in the LCOV rewriter.
+        """
         for artifact in self.artifactsHandler.get():
             if "jsvm" not in artifact:
                 continue
-
```
Comment (marco-c) [inline bot/code_coverage_bot/codecov.py]: Nit: don't remove newline

File: bot/code_coverage_bot/codecov.py [hunk 6]
```diff
@@ -146,27 +142,33 @@ def build_suites(self):
...
+            reports[(platform, suite)] = path
+
+        return reports
+
+    def upload_reports(self, reports):
+        """
+        Upload all provided covdir reports on GCP
+        """
+        for (platform, suite), path in reports.items():
+            report = json.load(open(path))
```
Comment (marco-c) [inline bot/code_coverage_bot/codecov.py]: You're loading as json here and then dumping as a string in `gcp`, you could instead pass the raw data and upload it directly.

File: bot/tests/test_artifacts.py (hunk 2 above)
Comment (La0) [reply bot/tests/test_artifacts.py]: Yes

File: bot/code_coverage_bot/uploader.py (hunk 3 above)
Comment (La0) [reply bot/code_coverage_bot/uploader.py]: If we remove that, we need a mechanism to detect the available platforms & suites in the backend

File: bot/code_coverage_bot/artifacts.py (hunk 1 above)
Comment (La0) [reply bot/code_coverage_bot/artifacts.py]: Counter-nit: `get_combinations` (as it's more generic to support future filters)

File: bot/code_coverage_bot/uploader.py (hunk 3 above)
Comment (marco-c) [reply bot/code_coverage_bot/uploader.py]: OK. Maybe we should have some kind of "partial" ingestion, but that's for the future.

Comment (La0) [reply bot/code_coverage_bot/uploader.py]: We could do a followup where the list of platforms/suites is simply `PUT` on the backend on `/v2/filters` to update the list of possible combinations.

File: bot/code_coverage_bot/codecov.py (hunk 6 above)
Comment (La0) [reply bot/code_coverage_bot/codecov.py]: Done.

**Pull request merged**